lib/filterx/func-sdata.c

```diff
@@ -0,0 +1,290 @@
+    {
+      const gchar *value = log_msg_get_value_name(msg->sdata[i], NULL);
+      gchar *at_sign = strchr(value, '@');
+      if (!at_sign)
+        continue;
+      contains = strncmp(at_sign+1, self->str_literal, self->str_literal_len) == 0;
+    }
+
+  return filterx_boolean_new(contains);
+}
+
+static void
+_free(FilterXExpr *s)
+{
+  FilterXFunctionIsSdataFromEnteprise *self = (FilterXFunctionIsSdataFromEnteprise *) s;
+
+  g_free(self->str_literal);
+  filterx_function_free_method(&self->super);
+}
+
+
+FilterXExpr *
+filterx_function_is_sdata_from_enterprise_new(const gchar *function_name, FilterXFunctionArgs *args, GError **error)
+{
+  FilterXFunctionIsSdataFromEnteprise *self = g_new0(FilterXFunctionIsSdataFromEnteprise, 1);
+  filterx_function_init_instance(&self->super, function_name);
+
+  if (!_extract_args(self, args, error) || !filterx_function_args_check(args, error))
+    goto error;
+  self->super.super.eval = _eval;
+  self->super.super.free_fn = _free;
+  filterx_function_args_free(args);
+  return &self->super.super;
+
+error:
+  filterx_function_args_free(args);
+  filterx_expr_unref(&self->super.super);
+  return NULL;
+}
+
+
+FilterXObject *
+filterx_simple_function_has_sdata(FilterXExpr *s, GPtrArray *args)
+{
+  if (args && args->len != 0)
+    {
+      filterx_simple_function_argument_error(s, "Incorrect number of arguments", FALSE);
+      return NULL;
+    }
+
+  FilterXEvalContext *context = filterx_eval_get_context();
+  LogMessage *msg = context->msgs[0];
+  return filterx_boolean_new(msg->num_sdata != 0);
+}
+
+static gboolean
+_should_create_new_dict(const gchar *previous_sd_id, gsize previous_sd_id_len, const gchar *current_sd_id,
+                        gsize current_sd_id_len)
+{
+  if (previous_sd_id_len != current_sd_id_len)
+    return TRUE;
+
+  return strncmp(previous_sd_id, current_sd_id, previous_sd_id_len) != 0;
+}
+
+
+static gboolean
+_extract_sd_components(const gchar *key, gssize key_len, const gchar **sd_id_start, gsize *sd_id_len,
+                       const gchar **sd_param_name_start, gsize *sd_param_name_len)
+{
+  /* skip '.SDATA.' */
+  if (key_len < SDATA_PREFIX_LEN)
+    return FALSE;
+
+  const gchar *last_dot = strrchr(key, '.');
+  if (!last_dot)
+    return FALSE;
+
+  *sd_id_start = key + SDATA_PREFIX_LEN;
+  *sd_id_len = last_dot - *sd_id_start;
+
+  *sd_param_name_start = last_dot + 1;
+  *sd_param_name_len = key + key_len - *sd_param_name_start;
+
+  return TRUE;
+}
+
+
+static gboolean
+_insert_into_dict(FilterXObject *dict, const gchar *key, gssize key_len, const gchar *value, gssize value_len)
+{
+  FilterXObject *fob_key = filterx_string_new(key, key_len);
+  FilterXObject *fob_value = filterx_string_new(value, value_len);
+  gboolean res = filterx_object_set_subscript(dict, fob_key, &fob_value);
+
+  filterx_object_unref(fob_key);
+  filterx_object_unref(fob_value);
+
+  return res;
+}
+
+static gboolean
+_generate(FilterXExprGenerator *s, FilterXObject *fillable)
```

Comment:
This function became really complex with the double loop and multiple iterator indexes being incremented and set manually. I now understand what is happening, but it could and should be a lot cleaner to be easily maintainable. Please refactor this function. I can help if you get stuck.